qiskit/pulse/transforms.py

```diff
@@ -31,92 +31,103 @@ def align_measures(schedules: Iterable[interfaces.ScheduleComponent],
         """Return the time needed to allow for readout discrimination calibration pulses."""
         max_calibration_duration = 0
         for qubits in inst_map.qubits_with_instruction(cal_gate):
             cmd = inst_map.get(cal_gate, qubits, np.pi, 0, np.pi)
             max_calibration_duration = max(cmd.duration, max_calibration_duration)
         return max_calibration_duration
 
-    if align_time is None and max_calibration_duration is None and inst_map is None:
-        raise exceptions.PulseError("Must provide a inst_map, an alignment time, "
-                                    "or a calibration duration.")
     if align_time is not None and align_time < 0:
         raise exceptions.PulseError("Align time cannot be negative.")
+
+    first_acquire_times = get_acquire_times(schedules)
+    max_acquire_times = [max(0, *times.values()) for times in first_acquire_times]
     if align_time is None:
-        align_time = calculate_align_time()
+        if max_calibration_duration is None:
+            if inst_map:
+                max_calibration_duration = get_max_calibration_duration(inst_map, cal_gate)
+            else:
+                max_calibration_duration = 0
+        align_time = max(max_calibration_duration, *max_acquire_times)
 
     # Shift acquires according to the new scheduled time
     new_schedules = []
-    for schedule in schedules:
+    for sched_idx, schedule in enumerate(schedules):
         new_schedule = Schedule(name=schedule.name)
-        acquired_channels = set()
-        measured_channels = set()
-
+        stop_time = schedule.stop_time
         for time, inst in schedule.instructions:
-            for chan in inst.channels:
-                if isinstance(chan, chans.MeasureChannel):
-                    if chan.index in measured_channels:
-                        raise exceptions.PulseError("Multiple measurements are "
-                                                    "not supported by this "
-                                                    "rescheduling pass.")
-                elif chan.index in acquired_channels:
-                    raise exceptions.PulseError("Pulse encountered on channel "
-                                                "{0} after acquire on "
-                                                "same channel.".format(chan.index))
-
-            if isinstance(inst, instructions.Acquire):
-                if time > align_time:
-                    warnings.warn("You provided an align_time which is scheduling an acquire "
-                                  "sooner than it was scheduled for in the original Schedule.")
-                new_schedule |= inst << align_time
-                acquired_channels.add(inst.channel.index)
-            elif isinstance(inst.channels[0], chans.MeasureChannel):
-                new_schedule |= inst << align_time
-                measured_channels.update({a.index for a in inst.channels})
+            measurement_channels = {
```

Comment:
We can filter the schedule by `schedule.channels` and insert the channel's schedule with the channel dependent time shift. I think this has smaller loops and perhaps faster (but this doesn't block merging because it will not be a matter with current program size).

Reply:
That would not be faster as it requires constructing a new schedule with `filter` and adds an extra inner loop.